httpbp/prometheus_test.go

```diff
@@ -0,0 +1,186 @@
+				},
+			},
+			"/error": {
+				Name:    "error",
+				Methods: []string{http.MethodGet},
+				Handle: func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+					return errors.New("test")
+				},
+			},
+		},
+		Middlewares: []Middleware{PrometheusServerMetrics(serverSlug)},
+	}
+
+	server, ts, err := NewTestBaseplateServer(args)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+
+	middleware := PrometheusClientMetrics(serverSlug)
+	client := &http.Client{
+		Transport: middleware(http.DefaultTransport),
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			serverLatency.Reset()
+			serverTotalRequests.Reset()
+			serverActiveRequests.Reset()
+			clientLatency.Reset()
+			clientTotalRequests.Reset()
+			clientActiveRequests.Reset()
+
+			serverSizeLabels := prometheus.Labels{
+				methodLabel:   tt.method,
+				successLabel:  tt.success,
+				endpointLabel: tt.endpoint,
+			}
+
+			serverTotalRequestLabels := prometheus.Labels{
+				methodLabel:   tt.method,
+				successLabel:  tt.success,
+				endpointLabel: tt.endpoint,
+				codeLabel:     tt.code,
+			}
+
+			serverActiveRequestLabels := prometheus.Labels{
+				methodLabel:   tt.method,
+				endpointLabel: tt.endpoint,
+			}
+
+			clientLatencyLabels := prometheus.Labels{
+				methodLabel:            tt.method,
+				successLabel:           tt.success,
+				endpointLabel:          tt.endpoint,
+				remoteServiceSlugLabel: serverSlug,
+			}
+
+			clientTotalRequestLabels := prometheus.Labels{
+				methodLabel:            tt.method,
+				successLabel:           tt.success,
+				endpointLabel:          tt.endpoint,
+				codeLabel:              tt.code,
+				remoteServiceSlugLabel: serverSlug,
+			}
+
+			clientActiveRequestLabels := prometheus.Labels{
+				methodLabel:            tt.method,
+				endpointLabel:          tt.endpoint,
+				remoteServiceSlugLabel: serverSlug,
+			}
+
+			defer promtest.NewPrometheusMetricTest(t, "server latency", serverLatency, serverSizeLabels).CheckExists()
+			defer promtest.NewPrometheusMetricTest(t, "server total requests", serverTotalRequests, serverTotalRequestLabels).CheckDelta(1)
+			defer promtest.NewPrometheusMetricTest(t, "server active requests", serverActiveRequests, serverActiveRequestLabels).CheckDelta(0)
+			defer promtest.NewPrometheusMetricTest(t, "server request size", serverRequestSize, serverSizeLabels).CheckDelta(float64(tt.size))
+			defer promtest.NewPrometheusMetricTest(t, "server response size", serverResponseSize, serverSizeLabels).CheckDelta(0)
+			defer promtest.NewPrometheusMetricTest(t, "client latency", clientLatency, clientLatencyLabels).CheckExists()
+			defer promtest.NewPrometheusMetricTest(t, "client total requests", clientTotalRequests, clientTotalRequestLabels).CheckDelta(1)
+			defer promtest.NewPrometheusMetricTest(t, "client active requests", clientActiveRequests, clientActiveRequestLabels).CheckDelta(0)
```

Comment:
Same question as above.

Reply:
i think originally the labels were getting modified later in the test so using a defer was helping use the correct labels. But now that doesn't seem to be occurring in this test.  I don't feel too strongly either way, but I have a slight preference keeping it the way it is now.